src/types/type_row.rs

```diff
@@ -0,0 +1,134 @@
+//! Rows of types, used for function signatures,
+//! designed to support efficient static allocation.
+
+use std::{
+    borrow::Cow,
+    fmt::{self, Display, Write},
+    ops::{Deref, DerefMut},
+};
+
+use crate::utils::display_list;
+
+/// Base trait for anything that can be put in a [TypeRow]
+pub trait TypeRowElem: std::fmt::Debug + Clone + 'static {}
+
+impl<T: Clone + std::fmt::Debug + 'static> TypeRowElem for T {}
```

Comment:
I think we can do without the `Debug` constraint here.
(Clone is needed for `[T]: ToOwned` in the `Cow`)

```suggestion
pub trait TypeRowElem: Clone + 'static {}

impl<T: Clone + 'static> TypeRowElem for T {}
```

Reply:
Neat! I didn't realize that. I presume `#[derive(Debug)]` automagically inserts the appropriate constraint i.e. `impl<T: Debug + ...> Debug for TypeRow<T>`